### DISCUSSION TITLE
Is there a way to config the maximum number of keys the server is allowed to return in the Range API call?

### DISCUSSION THREAD

**Discussion by wnwaynelee** [etcd-io/etcd#14808] [Q&A] [answered]
Hi,

The Range API (https://etcd.io/docs/v3.4/learning/api/#range) has a limit parameter the caller can specify the maximum number of keys to return.  The limit is type int64 and can be set to 0 for unlimited.

Is there a way to config an upper limit for the server response similar to –max-txn-ops and –max-request-bytes for request?  The intent is to prevent callers from stressing the server.

Thanks.

**Comment by lavacat** [comment, marked as answer]
No, there is no limit. To be precise it's 2147483647 bytes (MaxInt32 ~2gb). This was mentioned before in https://github.com/etcd-io/etcd/issues/14114. Might be worth adding this setting. cc @ahrtr 

> Currently the [maxSendBytes](https://github.com/etcd-io/etcd/blob/71bba3c761b0078c81c2b39781ec74853c458303/server/etcdserver/api/v3rpc/grpc.go#L36) is hard coded as math.MaxInt32. We may add a configuration item MaxSendMsgSize into [embed.Config](https://github.com/etcd-io/etcd/blob/71bba3c761b0078c81c2b39781ec74853c458303/server/embed/config.go#L144), and defaults to math.MaxInt32.

FYI, client-side you can use MaxCallRecvMsgSize to limit response.
